crates/node/src/catchup/blobs.rs

```diff
@@ -42,28 +36,21 @@ impl StreamFutures for ApplicationBlobChunkStream {
 
 impl AsyncRead for ApplicationBlobChunkStream {
     fn poll_read(
-        mut self: Pin<&mut Self>,
+        self: Pin<&mut Self>,
         cx: &mut StdContext<'_>,
         buf: &mut [u8],
     ) -> Poll<std::io::Result<usize>> {
-        if self.buffer.is_none() || self.offset >= self.buffer.as_ref().unwrap().len() {
-            match ready!(Pin::new(&mut self).poll_next(cx)) {
-                Some(chunk) => {
-                    self.buffer = Some(chunk.chunk);
-                    self.offset = 0;
-                }
-                None => return Poll::Ready(Ok(0)),
-            }
-        }
+        let this = self.get_mut();
 
-        if let Some(buffer) = &self.buffer {
-            let remaining = &buffer[self.offset..];
-            let len = remaining.len().min(buf.len());
-            buf[..len].copy_from_slice(&remaining[..len]);
-            self.offset += len;
-            Poll::Ready(Ok(len))
-        } else {
-            Poll::Ready(Ok(0))
+        match Pin::new(&mut this.receiver).poll_recv(cx) {
+            Poll::Ready(Some(chunk)) => {
+                let data = &chunk.chunk;
+                let len = data.len().min(buf.len());
+                buf[..len].copy_from_slice(&data[..len]);
```

Comment:
if the input buffer is less than what we read from the stream, this will truncate and discard the remaining bytes

we can use [`poll_fn`](https://docs.rs/futures-util/latest/futures_util/stream/fn.poll_fn.html) to poll [`poll_recv`](https://docs.rs/tokio/latest/tokio/sync/mpsc/struct.Receiver.html#method.poll_recv)

```rust
let stream = poll_fn(|cx| receiver.poll_recv(cx));
```

now we have a `Stream<Item = CatchupApplicationBlobChunk>` but to fix the issue mentioned above, we need to create a `BufReader` which requires the stream to be `AsyncRead`

so first, we can make the stream into a `Stream<Item = Result<Box<u8>, E>>` by:

```rust
let stream = stream.map(|chunk| {
    // check sequential_id, maybe return error?
    Ok(chunk.chunk)
});
```

and then we can make it into an `AsyncRead` with [`into_async_read`](https://docs.rs/futures-util/latest/futures_util/stream/trait.TryStreamExt.html#method.into_async_read)

```rust
let stream = stream.into_async_read();
```

At this point, we can pass it to [`BufReader::new`](https://docs.rs/futures-util/latest/futures_util/io/struct.BufReader.html#method.new)

```rust
let stream = BufReader::new(stream);
```

together it's about:

```rust
let stream = BufReader::new(
    poll_fn(|cx| receiver.poll_recv(cx))
        .map(|chunk| {
            // check sequential_id, maybe return error?
            Ok(chunk.chunk)
        })
        .into_async_read(),
);
```